### PR TITLE
Refactor join procedure and correct network calls

### PR DESCRIPTION
**Pull request by opomuc**
* Introduce new `net_connect` function to wrap net.box connections with correct fiber handling.
* Refactor join procedure
* Remove unused functions and legacy code
* Enhance error logging

### CODE REVIEW

File: shard.lua [hunk 1]
```diff
@@ -914,38 +924,45 @@ local function cluster_operation(func_name, ...)
...
-                     local conn = srd[i].conn
-                     local is_replica = i == 1
-                     result = conn[nb_call](conn, func_name, unpack(q_args), is_replica)[1][1]
-                 end)
-                 if not ok or not result then
-                     local msg = string.format(
+    for _, shard_set in ipairs(shards) do
+        for i, shard in ipairs(shard_set) do
+            if shard.id ~= id then
+                local ok, completed, err = pcall(function()
```
Comment (bigbes) [inline shard.lua]: Try to rewrite it not using pcall

File: shard.lua [hunk 2]
```diff
@@ -810,41 +831,31 @@ local function get_server_list()
...
-    if s_obj.conn == nil then
-        return false
-    end
-    if s_obj.conn:is_connected() then
-        maintenance[s_obj.id] = nil
+    -- check if connection exists
+    if s_obj.conn ~= nil and s_obj.conn:is_connected() then
+        if maintenance[s_obj.id] ~= nil then
+            maintenance[s_obj.id] = nil
+            log.info("Return '%s' from maintenance.", s_obj.uri)
```
Comment (bigbes) [inline shard.lua]: Get rid of dots in the end of the string. They don't have any meaning in the log file

Comment (bigbes) [reply shard.lua]: In all other places - too

File: shard.lua [hunk 3]
```diff
@@ -874,21 +885,20 @@ end
...
-                 shard_swap(j, i)
-                 to_join.ignore = false
-                 -- try to join replica
-                 return shard_connect(to_join)
-             end
-         end
+    for j, shard_set in ipairs(shards) do
+        for i, shard in ipairs(shard_set) do
+            if shard.id == id then
+                -- swap RO/RW shards
```
Comment (bigbes) [inline shard.lua]: Why're we swapping shards here?

Comment (bigbes) [reply shard.lua]: Add comment here, please

File: shard.lua [hunk 4]
```diff
@@ -914,38 +924,45 @@ local function cluster_operation(func_name, ...)
...
+                            "Server '%s' is offline.", shard.uri
+                        )
+                    end
+                    log.info(
+                        "Trying to '%s' to shard %d in node %s",
+                        func_name, shard.id, shard.uri
+                    )
+                    if q_args[2] == nil then
+                        local is_replica = (i ~= #shard_set)
+                        local result = shard.conn:call_16(func_name, unpack(q_args), is_replica)
```
[the remark this answers is not in the document]
Comment (bigbes) [reply shard.lua]: unpack will drop all arguments, except first one if used in the middle of the arguments

File: shard.lua [hunk 5]
```diff
@@ -914,38 +924,45 @@ local function cluster_operation(func_name, ...)
...
+                    )
+                    table.insert(jlog, msg)
+                    log.error(msg)
+                    all_ok = false
+                else
+                    local msg = string.format(
+                        "Operaion '%s' for shard %d in node '%s' applied",
+                        func_name, shard.id, shard.uri
+                    )
+                    table.insert(jlog, msg)
```
Comment (bigbes) [inline shard.lua]: please, rename `jlog` to something more appropriate

File: shard.lua [hunk 6]
```diff
@@ -1278,7 +1295,7 @@ end
 local function broadcast_call(task)
     local c = task.server.conn
```
Comment (bigbes) [inline shard.lua]: `c`? rename it, please

File: shard.lua [hunk 7]
```diff
@@ -810,41 +831,31 @@ local function get_server_list()
     return configuration.servers
 end
 
-local function is_shard_connected(uri, conn)
```
Comment (bigbes) [inline shard.lua]: Why you've deleted `is_shard_connected`? Where is it now or what is used instead of it.

Comment (opomuc) [reply shard.lua]: a better function `net_connect` is introduced. is_shard_connected was used only in `shard_connect`. Basically what I did is rewritten connect mechanism.

File: shard.lua [hunk 8]
```diff
@@ -1093,7 +1110,7 @@ local function direct_call(self, server, func_name, ...)
     local result = nil
     local status, reason = pcall(function(...)
         local conn = server.conn:timeout(REMOTE_TIMEOUT)
-        result = conn[nb_call](conn, func_name, ...)
+        result = conn:call_16(func_name, ...)
```
[the remark this answers is not in the document]
Comment (opomuc) [reply shard.lua]: I would be subtle to change calling convention of an API function. Lets leave it for another commit which will change every place where direct_call is used

File: shard.lua [hunk 9]
```diff
@@ -1278,7 +1295,7 @@ end
 local function broadcast_call(task)
     local c = task.server.conn
     local conn = c:timeout(REMOTE_TIMEOUT)
-    local tuples = conn[nb_call](conn, task.proc, unpack(task.args))
+    local tuples = conn:call_16(task.proc, unpack(task.args))
```
[the remark this answers is not in the document]
Comment (opomuc) [reply shard.lua]: see comment above

**Pull request merged**